### PR TITLE
Added guild templates

### PR DESCRIPTION
**Pull request by AnotherZane**

### CODE REVIEW

File: src/Disqord.Core/Models/GuildTemplateJsonModel.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+        [JsonProperty("updated_at")]
+        public DateTimeOffset UpdatedAt;
+
+        [JsonProperty("source_guild_id")]
+        public Snowflake SourceGuildId;
+
+        // TODO: Figure out what to do here since this object does not have properties such as id, etc.
+        // [JsonProperty("serialized_source_guild")]
+        // public GuildJsonModel SerializedSourceGuild;
```
Comment (AnotherZane) [inline src/Disqord.Core/Models/GuildTemplateJsonModel.cs:37]: Probably a good idea to keep this open until a solution to this is found.

Comment (Quahu) [reply src/Disqord.Core/Models/GuildTemplateJsonModel.cs:37]: I believe simply exposing the `IJsonObject`, even in the public API/entities, might just be the way to go about it. It's not like the user can do anything useful with the object programmatically anyways, and they can view it easily by just calling `ToString()` which will get them the JSON representation.

Comment (AnotherZane) [reply src/Disqord.Core/Models/GuildTemplateJsonModel.cs]: I have made it an `IJsonObject` in the recent commits. Maybe we could possibly look into creating an `IGuild` from this by using the `SourceGuildId` as the Id?

File: src/Disqord.Core/Entities/Core/Guild/IGuildTemplate.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        
+        /// <summary>
+        ///     Gets when this template was updated.
+        /// </summary>
+        DateTimeOffset UpdatedAt { get; }
+        
+        /// <summary>
+        ///     Gets the serialized source guild of the template.
+        /// </summary>
+        IJsonObject SerializedSourceGuild { get; }
```
Comment (Quahu) [inline src/Disqord.Core/Entities/Core/Guild/IGuildTemplate.cs:50]: This should be just `SerializedGuild`.

File: src/Disqord.Rest/Extensions/Entity/RestEntityExtensions.Guild.cs [hunk 3]
```diff
@@ -282,5 +282,44 @@ public static Task<IReadOnlyList<IWebhook>> FetchWebhooksAsync(this IGuild guild
...
         }
+        
+        /*
+         * Template
+         */
+        public static Task<IGuildTemplate> FetchTemplateAsync(this IGuild guild, string templateCode, IRestRequestOptions options = null)
+        {
+            var client = guild.GetRestClient();
+            return client.FetchTemplateAsync(templateCode, options);
+        }
```
Comment (Quahu) [inline src/Disqord.Rest/Extensions/Entity/RestEntityExtensions.Guild.cs:293]: This is a global method, it shouldn't be on the guild object.

File: src/Disqord.Core/Entities/Core/Guild/IGuildTemplate.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        /// </summary>
+        string Description { get; }
+        
+        /// <summary>
+        ///     Gets the usage count of the template.
+        /// </summary>
+        int UsageCount { get; }
+        
+        /// <summary>
+        ///     Gets the creator ID of the template.
```
Comment (Quahu) [inline src/Disqord.Core/Entities/Core/Guild/IGuildTemplate.cs:28]: "Gets the ID of the creator of this template."

File: src/Disqord.Core/Entities/Core/Guild/IGuildTemplate.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        
+        /// <summary>
+        ///     Gets the description of the template.
+        /// </summary>
+        string Description { get; }
+        
+        /// <summary>
+        ///     Gets the usage count of the template.
+        /// </summary>
+        int UsageCount { get; }
```
Comment (Quahu) [inline src/Disqord.Core/Entities/Core/Guild/IGuildTemplate.cs:25]: I say we call this `Uses`, to match how the property is called on invites.

**Review by Quahu — CHANGES_REQUESTED** [review]
Could you also add the entity extensions for manipulating the template entities? `template.DeleteAsync()` and so on.

**Review by Quahu — APPROVED** [review]

**Pull request merged**